src/ralph/virtual/management/commands/openstack_sync.py

```diff
@@ -0,0 +1,532 @@
+            projects[project.id] = {
+                'name': project.name,
+                'tag':  site['tag'],
+                'servers': {},
+            }
+
+        if len(projects) == 0:
+            raise NotImplementedError('Got an empty list of projects from '
+                                      '{}'.format(site['auth_url']))
+        return projects
+
+    def _run_openstack_query(self):
+        for site in settings.OPENSTACK_INSTANCES:
+            nt = self._get_novaclient_connection(site)
+            tmp = self.openstack_projects.copy()
+            tmp.update(self._get_projects_map(site))
+            for server in self._get_servers_list(nt, site):
+                project_id = server['_info']['tenant_id']
+                host_id = server['_info']['hostId']
+                new_server = {
+                    'hostname': server['_info']['name'],
+                    'flavor_id': server['_info']['flavor']['id'],
+                    'tag': site['tag'],
+                    'ips': [],
+                    'created': server['_info']['created'],
+                    'modified': server['_info']['updated']
+                }
+                for zone in server['_info']['addresses']:
+                    for ip in server['_info']['addresses'][zone]:
+                        new_server['ips'].append(ip['addr'])
+                tmp[project_id]['servers'][host_id] = new_server
+            self.openstack_projects = tmp
+
+            for flavor in self._get_flavors_list(nt, site):
+                flavor_id = flavor['_info']['id']
+                new_flavor = {
+                    'name': flavor['_info']['name'],
+                    'cores': flavor['_info']['vcpus'],
+                    'ram': flavor['_info']['ram'],
+                    'disk': flavor['_info']['disk'],
+                }
+                self.openstack_flavors[flavor_id] = new_flavor
+
+    def _get_cloud_provider(self):
+        """Get or create cloud provider object"""
+        try:
+            self.cloud_provider = CloudProvider.objects.get(name='openstack')
+        except ObjectDoesNotExist:
+            self.cloud_provider = CloudProvider(name='openstack')
+            self._save_object(self.cloud_provider,
+                              'Add openstack CloudProvider')
+
+    def _get_ralph_data(self):
+        """Get configuration from ralph DB"""
+        self.ralph_projects = {}
+        self.ralph_flavors = {}
+        for project in CloudProject.objects.filter(
+            cloudprovider=self.cloud_provider
+        ):
+            project_id = project.project_id
+            self.ralph_projects[project_id] = {
+                'name': project.name,
+                'servers': {},
+                'tags': project.tags.names()
+            }
+            try:
+                self.ralph_projects[project_id][
+                    'cloudprovider'] = project.cloudprovider
+            except ObjectDoesNotExist:
```

Comment:
when could this error occur? Since `cloudprovider` is mandatory field for `CloudProject`, this try-except is useless here